contracts/L1/gateway/L1Migrator.sol

```diff
@@ -0,0 +1,306 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {L1ArbitrumMessenger} from "./L1ArbitrumMessenger.sol";
+import {IMigrator} from "../../interfaces/IMigrator.sol";
+import {ECDSA} from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import {EIP712} from "@openzeppelin/contracts/utils/cryptography/draft-EIP712.sol";
+
+interface IBondingManager {
+    function isRegisteredTranscoder(address _addr) external view returns (bool);
+    function pendingStake(address _addr, uint256 _endRound) external view returns (uint256);
+    function pendingFees(address _addr, uint256 _endRound) external view returns (uint256);
+    function getDelegator(address _addr) external view returns (
+        uint256 bondedAmount,
+        uint256 fees,
+        address delegateAddress,
+        uint256 delegatedAmount,
+        uint256 startRound,
+        uint256 lastClaimRound,
+        uint256 nextUnbondingLockId
+    );
+    function getDelegatorUnbondingLock(
+        address _addr,
+        uint256 _unbondingLockId
+    ) external view returns (uint256 amount, uint256 withdrawRound);
+}
+
+interface ITicketBroker {
+    struct Sender {
+        uint256 deposit;
+        uint256 withdrawRound;
+    }
+
+    struct ReserveInfo {
+        uint256 fundsRemaining;
+        uint256 claimedInCurrentRound;
+    }
+
+    function getSenderInfo(address _addr)
+        external
+        view
+        returns (
+            Sender memory sender,
+            ReserveInfo memory reserve
+        );
+}
+
+interface IL2Migrator is IMigrator {
+    function finalizeMigrateDelegator(MigrateDelegatorParams memory _params) external;
+    function finalizeMigrateUnbondingLocks(MigrateUnbondingLocksParams memory _params) external;
+    function finalizeMigrateSender(MigrateSenderParams memory _params) external;
+}
+
+contract L1Migrator is L1ArbitrumMessenger, IMigrator, EIP712 {
+    address public immutable bondingManagerAddr;
+    address public immutable ticketBrokerAddr;
+    address public immutable l2MigratorAddr;
+
+    event MigrateDelegatorInitiated(
+        uint256 indexed seqNo,
+        MigrateDelegatorParams params
+    );
+
+    event MigrateUnbondingLocksInitiated(
+        uint256 indexed seqNo,
+        MigrateUnbondingLocksParams params
+    );
+
+    event MigrateSenderInitiated(
+        uint256 indexed seqNo,
+        MigrateSenderParams params
+    );
+
+    bytes32 private constant MIGRATE_DELEGATOR_TYPE_HASH = 
+        keccak256("MigrateDelegator(address l1Addr,address l2Addr)");
+
+    bytes32 private constant MIGRATE_UNBONDING_LOCKS_TYPE_HASH = 
+        keccak256("MigrateUnbondingLocks(address l1Addr,address l2Addr,uint256[] unbondingLockIds)");
+
+    bytes32 private constant MIGRATE_SENDER_TYPE_HASH = 
+        keccak256("MigrateSender(address l1Addr,address l2Addr)");
+
+    constructor(
+        address _inbox,
+        address _bondingManagerAddr,
+        address _ticketBrokerAddr,
+        address _l2MigratorAddr
```

Comment:
Since we are not using deterministic deployment as of now there would be a cyclic dependency for both the migrators as one depends on the other. 

Right now there is a function in the L2 migrator to set L1 migrator address which resolves this if L2 migrator is deployed first but flagging this for future